libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1746 @@
+  {}
+
+  _LIBCUDACXX_INLINE_VISIBILITY ~__inplace_vector_destruct_base() noexcept
+  {
+    _Tp* __begin = _CUDA_VSTD::launder(reinterpret_cast<_Tp*>(__elems_));
+    _CUDA_VSTD::__destroy(__begin, __begin + __size_);
+  }
+};
+
+template <class _Tp, size_t _Capacity>
+struct __inplace_vector_destruct_base<_Tp, _Capacity, __inplace_vector_specialization::__trivial_destructible>
+{
+  using size_type   = size_t;
+  using __size_type = __inplace_vector_size_type<_Capacity>;
+
+  alignas(_Tp) unsigned char __elems_[_Capacity * sizeof(_Tp)] = {};
+  __size_type __size_{0};
+
+  // [containers.sequences.inplace.vector.cons], construct/copy/destroy
+  constexpr __inplace_vector_destruct_base() = default;
+
+  _LIBCUDACXX_INLINE_VISIBILITY __inplace_vector_destruct_base(const __inplace_vector_destruct_base& __other) noexcept(
+    _CCCL_TRAIT(is_nothrow_copy_constructible, _Tp))
+      : __elems_()
+      , __size_(__other.size())
+  {}
+
+  _LIBCUDACXX_INLINE_VISIBILITY __inplace_vector_destruct_base(__inplace_vector_destruct_base&& __other) noexcept(
+    _CCCL_TRAIT(is_nothrow_move_constructible, _Tp))
+      : __elems_()
+      , __size_(__other.size())
+  {}
+
+  // Everything is done in __inplace_vector_base
+  _LIBCUDACXX_INLINE_VISIBILITY __inplace_vector_destruct_base&
+  operator=(const __inplace_vector_destruct_base&) noexcept(
+    _CCCL_TRAIT(is_nothrow_copy_constructible, _Tp) && _CCCL_TRAIT(is_nothrow_copy_assignable, _Tp))
+  {}
+  _LIBCUDACXX_INLINE_VISIBILITY __inplace_vector_destruct_base& operator=(__inplace_vector_destruct_base&&) noexcept(
+    _CCCL_TRAIT(is_nothrow_move_constructible, _Tp) && _CCCL_TRAIT(is_nothrow_move_assignable, _Tp))
+  {}
+};
+
+template <class _Tp,
+          size_t _Capacity,
+          __inplace_vector_specialization _Spec = __select_inplace_vector_specialization<_Tp, _Capacity>()>
+struct __inplace_vector_base : __inplace_vector_destruct_base<_Tp, _Capacity, _Spec>
+{
+  using size_type      = size_t;
+  using __size_type    = __inplace_vector_size_type<_Capacity>;
+  using reference      = _Tp&;
+  using iterator       = _Tp*;
+  using const_iterator = const _Tp*;
+
+  // [containers.sequences.inplace.vector.cons], construct/copy/destroy
+  constexpr __inplace_vector_base() = default;
+  _LIBCUDACXX_INLINE_VISIBILITY
+  __inplace_vector_base(const __inplace_vector_base& __other) noexcept(_CCCL_TRAIT(is_nothrow_copy_constructible, _Tp))
+      : __inplace_vector_destruct_base<_Tp, _Capacity, _Spec>()
+  {
+    __uninitialized_copy(__other.begin(), __other.end(), begin());
+  }
+  _LIBCUDACXX_INLINE_VISIBILITY
+  __inplace_vector_base(__inplace_vector_base&& __other) noexcept(_CCCL_TRAIT(is_nothrow_move_constructible, _Tp))
+      : __inplace_vector_destruct_base<_Tp, _Capacity, _Spec>()
+  {
+    __uninitialized_move(__other.begin(), __other.end(), begin());
+    __other.__destroy(__other.begin(), __other.end());
+  }
```

Comment:
Its a bit tricky. 

The defaulted copy / move constructors in the base classes would actually copy data around. However, that is not what we want because that data will be invalid when the type is not trivially copy / move constructible.

So we do nothing in the base class and all the work here. 

Also note that otherwise we would need to move all the functions like `__uninitialized_move` into the base class which duplicates a lot of code

Reply:
Please add a comment to the special members of `__inplace_vector_destruct_base` why we do that.